pkg/informer/informer.go

```diff
@@ -491,3 +562,17 @@ func (d *DynamicDiscoverySharedInformerFactory) calculateInformersLockHeld(lates
 
 	return toAdd, toRemove
 }
+
+// Subscribe registers ch for informer/discovery change notifications.
+func (d *DynamicDiscoverySharedInformerFactory) Subscribe(id string, ch chan<- struct{}) {
```

Comment:
If we let users pass us their end of the channel, we need to handle being blocked on sending a notification. Alternatively, this can *return* them a channel (and we can choose to make it buffered, but then also need to handle the case where we can't notify them within some timeout)

Reply:
this seems brittle. This means we drop events as soon as ch has no capacity at the moment. We cannot do that.

IMO we should not use channels at all. It's a poor type for APIs. Make this a function call. Document that the consumer must make sure to return quickly and not block. The consumer will know how to make that work. It has the context to do that. Here we don't.

Reply:
FWIW this is how watch server works, so while brittle it's likely workable?

Reply:
I would prefer to keep this as-is, for a few of reasons:

1. The event dropping is on purpose. I want to deduplicate notifications, and this is easily achieved with a buffer of size 1.
2. The consumer in this case is not guaranteed to return quickly. It is doing a WaitForCacheSync. It currently waits "forever" (until either server shutdown, or another notification comes in (coming in next push))
3. I need to be able to cancel the consumer's WaitForCacheSync, which I can't easily do with a single function call

Reply:
> The event dropping is on purpose. I want to deduplicate notifications, and this is easily achieved with a buffer of size 1.

It's not the duty of the informer to handle that.

> The consumer in this case is not guaranteed to return quickly. It is doing a WaitForCacheSync. It currently waits "forever" (until either server shutdown, or another notification comes in (coming in next push))

It is not the duty of the informer to know what the subscribers do or don't.

> I need to be able to cancel the consumer's WaitForCacheSync, which I can't easily do with a single function call

Why?

Reply:
> FWIW this is how watch server works, so while brittle it's likely workable?

It's not about whether it technically works or not in other context. It's about making complexity local. The current interface does not archive that.

Reply:
This also support my logic: the go doc here says nothing about dropping event. I as subscriber would be highly surprised if I lose these events just because I am a little too slow. This asks for races that are super hard to debug. All this wouldn't be the case with a simple func based callback and local deduplication/dropping logic in the consumer side.